### PR TITLE
Research for WM + battle texture replacement

### PR DESCRIPTION
**Pull request by myst6re**
## Summary

Add partial support to battle textures
+ Fix texture holes in Worldmap

### ACKs

- [X] I have updated the [Changelog.md](https://github.com/julianxhokaxhiu/FFNx/blob/master/Changelog.md) file
- [X] I did test my code on FF7
- [X] I did test my code on FF8

https://imgsli.com/MTg5NjQ2

### CODE REVIEW

File: src/ff8/vram.cpp [hunk 1]
```diff
@@ -575,20 +593,284 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
 	texturePacker.setTexture(texture_name, pos_and_size[0], pos_and_size[1], pos_and_size[2], pos_and_size[3] * 2, bpp, false);
 }
 
-DWORD *create_graphics_object_load_texture_call2(int a1, int a2, char *path, void *data, ff8_game_obj *game_object)
+Stage stage;
+
+int16_t ff8_battle_open_and_read_file(int fileId, void *data, int a3, int callback)
+{
+	char **battle_filenames = (char **)0xB84CCC;
```
Comment (myst6re) [inline src/ff8/vram.cpp:597]: The offset is hard to get

File: src/ff8/vram.cpp [hunk 2]
```diff
@@ -575,20 +593,284 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+	snprintf(battle_texture_name, sizeof(battle_texture_name), "battle/%s", battle_filenames[fileId]);
+
+	return ((int16_t(*)(int,void*,int,int))0x482600)(fileId, data, a3, callback);
```
Comment (myst6re) [inline src/ff8/vram.cpp:600]: The offset is hard to get

File: src/ff8/vram.cpp [hunk 3]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+	uint8_t data_segment;
+	uint8_t padding;
+};
+
+int data_segment = 0;
+int current_segment = -1;
+int current_block = -1;
+int current_polygon = -1;
+
+wmx_linked_list_element *worldmap_get_wm_obj_struc_sub_553B30(__int16 segment, wmx_linked_list_element *wmx_object)
```
Comment (myst6re) [inline src/ff8/vram.cpp:676]: Change this name

File: src/ff8/vram.cpp [hunk 4]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+{
+	wmx_linked_list_element *ret = ((wmx_linked_list_element*(*)(__int16,wmx_linked_list_element*))0x553B30)(segment, wmx_object);
```
Comment (myst6re) [inline src/ff8/vram.cpp:678]: Set the correct offset

File: src/ff8/vram.cpp [hunk 5]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
 	return ret;
 }
 
+void *current_polygon_condition_old = nullptr;
+bool old_dword_2045968_value = false;
+uint8_t *current_block_data_start = nullptr;
+
+void worldmap_fog_common(uint8_t *block, int a2, void *out)
+{
+	uint8_t *wmx_data_off_C75DB4 = *(uint8_t **)0xC75DB4;
```
Comment (myst6re) [inline src/ff8/vram.cpp:694]: Set the correct offset

File: src/ff8/vram.cpp [hunk 6]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+	uint8_t *segment = wmx_data_off_C75DB4 + data_segment * 0x9000;
```
Comment (myst6re) [inline src/ff8/vram.cpp:695]: Set the correct offset

File: src/ff8/vram.cpp [hunk 7]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+	uint32_t block_offset = uint32_t(block - segment);
+	uint32_t dword_20359D8 = *(uint32_t *)0x20359D8;
```
Comment (myst6re) [inline src/ff8/vram.cpp:697]: Set the correct offset

File: src/ff8/vram.cpp [hunk 8]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+			current_block = i;
+
+			break;
+		}
+	}
+}
+
+void worldmap_fog_common_leave()
+{
+	void **current_polygon_condition_dword_2045964 = (void **)0x2045964;
```
Comment (myst6re) [inline src/ff8/vram.cpp:713]: Set the correct offset

File: src/ff8/vram.cpp [hunk 9]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+	bool *dword_2045968 = (bool *)0x2045968;
```
Comment (myst6re) [inline src/ff8/vram.cpp:714]: Set the correct offset

File: src/ff8/vram.cpp [hunk 10]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+
+	*dword_2045968 = old_dword_2045968_value;
+	*current_polygon_condition_dword_2045964 = current_polygon_condition_old;
+}
+
+void worldmap_fog_sub_551AC0(uint8_t *block, int a2, void *out)
+{
+	worldmap_fog_common(block, a2, out);
+
+	((void(*)(uint8_t*,int,void*))0x551AC0)(block, a2, out);
```
Comment (myst6re) [inline src/ff8/vram.cpp:724]: Set the correct offset

File: src/ff8/vram.cpp [hunk 11]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+
+	worldmap_fog_common_leave();
+}
+
+void worldmap_fog_sub_552480(uint8_t *block, int a2, void *out)
+{
+	worldmap_fog_common(block, a2, out);
+
+	((void(*)(uint8_t*,int,void*))0x552480)(block, a2, out);
```
Comment (myst6re) [inline src/ff8/vram.cpp:733]: Set the correct offset

File: src/ff8/vram.cpp [hunk 12]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+	if (!old_dword_2045968_value) {
+		return true;
+	}
+
+	return ((bool(*)(int))current_polygon_condition_old)(polygon_id);
+}
+
+void sub_45DF10(int a1, int a2, int a3)
+{
+	void **current_polygon_condition_dword_2045964 = (void **)0x2045964;
```
Comment (myst6re) [inline src/ff8/vram.cpp:756]: Set the correct offset

File: src/ff8/vram.cpp [hunk 13]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+	bool *dword_2045968 = (bool *)0x2045968;
```
Comment (myst6re) [inline src/ff8/vram.cpp:757]: Set the correct offset

File: src/ff8/vram.cpp [hunk 14]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+
+	old_dword_2045968_value = *dword_2045968;
+
+	if (!old_dword_2045968_value) {
+		*dword_2045968 = true;
+		current_polygon_condition_old = *current_polygon_condition_dword_2045964;
+		*current_polygon_condition_dword_2045964 = current_polygon_condition;
+	}
+
+	((void(*)(int,int,int))0x45DF10)(a1, a2, a3);
```
Comment (myst6re) [inline src/ff8/vram.cpp:767]: Set the correct offset

File: src/ff8/vram.cpp [hunk 15]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+};
+
+void enrich_tex_coords_sub_45E390f(struc_33 *a1)
+{
+	MapBlockPolygon *polygon = (MapBlockPolygon *)(current_block_data_start + 4 + current_polygon * sizeof(MapBlockPolygon));
+	a1->field_26 = ((polygon->pos[0].x & 1) << 0) | ((polygon->pos[0].y & 1) << 1)
+		| ((polygon->pos[1].x & 1) << 2) | ((polygon->pos[1].y & 1) << 3)
+		| ((polygon->pos[2].x & 1) << 4) | ((polygon->pos[2].y & 1) << 5);
+
+	((void(*)(struc_33*))0x45E390)(a1);
```
Comment (myst6re) [inline src/ff8/vram.cpp:793]: Set the correct offset

File: src/ff8/vram.cpp [hunk 16]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+
+float *maybe_hundred_bak = nullptr;
+
+void ssigpu_callback_vertex_common(struc_33 *a1, uint32_t cb)
+{
+	uint8_t lost_bits = a1->field_26;
+
+	((void(*)(struc_33*))cb)(a1);
+
+	int dword_1CA8520 = *(int *)0x1CA8520;
```
Comment (myst6re) [inline src/ff8/vram.cpp:804]: Set the correct offset

**Comment by julianxhokaxhiu** [comment]
Is it safe to assume that this patch fixed also the UV coords for WM? So in case we can delete [the hext files for FF8](https://github.com/julianxhokaxhiu/FFNx/tree/master/misc/hext/ff8) directly in this PR. What do you think?

File: src/ff8/vram.cpp [hunk 17]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+		maybe_hundred_bak[22] = double(uint32_t(a1->tex_coord_c.x) * 2 + ((lost_bits >> 4) & 1)) / 511.999;
+		maybe_hundred_bak[23] = double(uint32_t(a1->tex_coord_c.y) * 2 + ((lost_bits >> 5) & 1)) / 511.999;
+
+		maybe_hundred_bak = nullptr;
+	}
+}
+
+void ssigpu_callback_sub_4617E0(struc_33 *a1)
+{
+	ssigpu_callback_vertex_common(a1, 0x4617E0);
```
Comment (myst6re) [inline src/ff8/vram.cpp:820]: Set the correct offset

File: src/ff8/vram.cpp [hunk 18]
```diff
@@ -575,20 +590,260 @@ void ff8_field_effects_upload_vram1(int16_t *pos_and_size, uint8_t *texture_buff
...
+}
+
+void ssigpu_callback_sub_461A80(struc_33 *a1)
+{
+	ssigpu_callback_vertex_common(a1, 0x461A80);
```
Comment (myst6re) [inline src/ff8/vram.cpp:825]: Set the correct offset

**Comment by myst6re** [comment]
> Is it safe to assume that this patch fixed also the UV coords for WM? So in case we can delete [the hext files for FF8](https://github.com/julianxhokaxhiu/FFNx/tree/master/misc/hext/ff8) directly in this PR. What do you think?

Nope, let me explain:

The Maki patch happen right after the parsing of file data, it prevent the game to alter the UVs (the game does it to use bilinear filtering on textures)

Later the game filter which vertices are currently displayed on the screen and produce a big array of GPU commands with various things like vertices. But it alters the UVs again in some cases in this big array. Later again, the game reads the big array and transform the UVs into floats. That's where my patch fixes things

So in short:

1. WMX parsing
2. UVs modified (2 added/substracted) <-- Maki patch to remove this
3. Vertices targeting
4. UVs modified (targeted to the right position in the texture chunk of the VRAM) <-- My PR stores the missing bits caused by this modification
5. Big array of GPU commands created with the targeted vertices
6. Execution of GPU commands
7. UVs converted to float <-- My PR restores the missing bits here for more accurate floating values

**Comment by myst6re** [comment]
I can integrate the maki patch into FFNx and make it an option in FFNx.toml if it is needed

**Comment by julianxhokaxhiu** [comment]
> I can integrate the maki patch into FFNx and make it an option in FFNx.toml if it is needed

No worries, it's not a high prio, we can include it in code any time :) Thanks for working on this in the meantime

**Review by julianxhokaxhiu — APPROVED** [review]

**Comment by julianxhokaxhiu** [comment]
Fantastic work there, thank you!
